hail/src/main/scala/is/hail/rvd/RVDPartitioner.scala

```diff
@@ -11,55 +11,91 @@ class RVDPartitioner(
   val kType: TStruct,
   // rangeBounds: Array[Interval[kType]]
   // rangeBounds is interval containing all keys within a partition
-  val rangeBounds: IndexedSeq[Interval],
+  val rangeBounds: Array[Interval],
   allowedOverlap: Int
 ) {
+  def this(
+    kType: TStruct,
+    rangeBounds: IndexedSeq[Interval],
+    allowedOverlap: Int
+  ) = this(kType, rangeBounds.toArray, kType.size)
+
   def this(
     kType: TStruct,
     rangeBounds: IndexedSeq[Interval]
-  ) = this(kType, rangeBounds, kType.size)
+  ) = this(kType, rangeBounds.toArray, kType.size)
 
   def this(
     partitionKey: Array[String],
     kType: TStruct,
     rangeBounds: IndexedSeq[Interval]
-  ) = this(kType, rangeBounds, math.max(partitionKey.length - 1, 0))
+  ) = this(kType, rangeBounds.toArray, math.max(partitionKey.length - 1, 0))
 
   def this(
     partitionKey: Option[Int],
     kType: TStruct,
     rangeBounds: IndexedSeq[Interval]
-  ) = this(kType, rangeBounds, partitionKey.map(_ - 1).getOrElse(kType.size))
+  ) = this(kType, rangeBounds.toArray, partitionKey.map(_ - 1).getOrElse(kType.size))
 
   require(rangeBounds.forall { case Interval(l, r, _, _) =>
     kType.relaxedTypeCheck(l) && kType.relaxedTypeCheck(r)
   })
   require(allowedOverlap >= 0 && allowedOverlap <= kType.size)
   require(RVDPartitioner.isValid(kType, rangeBounds, allowedOverlap))
 
+  val kord: ExtendedOrdering = kType.ordering
+  val eord: ExtendedOrdering = kord.intervalEndpointOrdering
+  val qord: (Any, Any) => Boolean = Interval.quasiordering(kord)
+  val scalaEOrd: Ordering[IntervalEndpoint] =
+    kord.intervalEndpointOrdering.toOrdering.asInstanceOf[Ordering[IntervalEndpoint]]
+
+  def range: Option[Interval] =
+    if (rangeBounds.isEmpty)
+      None
+    else
+      Some(Interval(rangeBounds.head.left, rangeBounds.last.right))
```

Comment:
I believe `.last` is O(N). there's certainly no guarantee of O(1)

Reply:
This is on an array, being implicitly converted to an `IndexedSeqOptimized`. The definition is just `if (length > 0) this(length - 1) else super.last`. I know Scala has issues, but I have to believe it's safe to assume that accessing an index into an array (which `last` is clearly syntactic sugar for) is O(1).